### PR TITLE
Enable enforcing defaults

### PR DESCRIPTION
**Pull request by RobbeSneyders**
This PR is a proposal to enforce defaults in json bodies.

I'm not sure which way to go here
- I think connexion should offer the functionality to fill in the defaults, as this helps enforce the contract. If a value is not provided, the application gets the default. We should probably offer this using a simple `default` flag instead of the custom validator now needed.
- I don't like that validation adapts the body. Validation is a read-only operation and it would be nice to keep it that way.
- Splitting enforcing defaults from validation would be less efficient since we'd need to traverse the body multiple times.

If we do want to go this way, we should probably implement a similar fix for form data, as we also use json validators there.

### CODE REVIEW

**Comment by RobbeSneyders** [comment]
I updated the PR with a solution that I think is a good middle ground.

The standard `JsonRequestBodyValidator` does not adapt the body. Instead, I added a `DefaultsJsonRequestBodyValidator` which does. The user can easily activate enforcing default by passing this Validator in a `validator_map`.

I would have liked connexion to have an `enforce_defaults` flag, but it would be hard to make this compatible with the `validator_map` argument.

File: connexion/apps/asynchronous.py [hunk 1]
```diff
@@ -63,7 +63,8 @@ def fn(self) -> t.Callable:
     async def __call__(
         self, scope: Scope, receive: Receive, send: Send
     ) -> StarletteResponse:
-        return await self.fn(scope=scope, receive=receive, send=send)
```
Comment (RobbeSneyders) [inline connexion/apps/asynchronous.py:66]: Just a fix to get the `AsyncApp` working correctly. Not related to enforcing defaults, but necessary to make the example work.

File: connexion/apps/asynchronous.py [hunk 2]
```diff
@@ -78,15 +79,17 @@ def make_operation(self, operation: AbstractOperation) -> AsyncOperation:
...
     def __init__(self) -> None:
         self.apis: t.Dict[str, AsyncApi] = {}
         self.operations: t.Dict[str, AsyncOperation] = {}
         self.router = Router()
         super().__init__(self.router)
 
     def add_api(self, *args, **kwargs):
-        api = AsyncApi(*args, **kwargs)
-        self.apis[api.base_path] = api
+        api = super().add_api(*args, **kwargs)
```
Comment (RobbeSneyders) [inline connexion/apps/asynchronous.py:92]: Just a fix to get the `AsyncApp` working correctly. Not related to enforcing defaults, but necessary to make the example work.

File: connexion/decorators/main.py [hunk 3]
```diff
@@ -166,10 +166,7 @@ async def wrapper(*args, **kwargs):
                 uri_parser=self.uri_parser, scope=scope, receive=receive
             )
             decorated_function = self.decorate(function)
-            response = decorated_function(request)
-            while asyncio.iscoroutine(response):
-                response = await response
-            return response
+            return await decorated_function(request)
```
Comment (RobbeSneyders) [inline connexion/decorators/main.py:169]: Just a fix to get the `AsyncApp` working correctly. Not related to enforcing defaults, but necessary to make the example work.

File: connexion/middleware/abstract.py [hunk 4]
```diff
@@ -244,11 +244,10 @@ def __init__(self, app: ASGIApp) -> None:
         self.app = app
         self.apis: t.Dict[str, API] = {}
 
-    def add_api(
-        self, specification: t.Union[pathlib.Path, str, dict], **kwargs
-    ) -> None:
+    def add_api(self, specification: t.Union[pathlib.Path, str, dict], **kwargs) -> API:
         api = self.api_cls(specification, next_app=self.app, **kwargs)
         self.apis[api.base_path] = api
+        return api
```
Comment (RobbeSneyders) [inline connexion/middleware/abstract.py:250]: Just a fix to get the `AsyncApp` working correctly. Not related to enforcing defaults, but necessary to make the example work.

File: examples/sqlalchemy/spec/openapi.yaml [hunk 5]
```diff
@@ -62,10 +62,10 @@ paths:
         '201':
           description: New pet created
       requestBody:
+        x-body-name: pet
```
Comment (RobbeSneyders) [inline examples/sqlalchemy/spec/openapi.yaml:65]: Unrelated fix

File: connexion/validators/json.py [hunk 6]
```diff
@@ -52,30 +52,107 @@ def validate(self, body: dict):
...
+            validator_class, {"properties": set_defaults}
+        )
+
+    async def read_body(self) -> t.Tuple[str, int]:
+        """Read the body from the receive channel.
+
+        :return: A tuple (body, max_length) where max_length is the length of the larges message.
+        """
+        more_body = True
+        max_length = 256000
```
Comment (Ruwann) [inline connexion/validators/json.py:114]: Where does this number come from? I'm not aware of an ASGI limit

File: connexion/validators/json.py [hunk 7]
```diff
@@ -52,30 +52,107 @@ def validate(self, body: dict):
...
+            yield from validate_properties(validator, properties, instance, schema)
+
+        return jsonschema.validators.extend(
+            validator_class, {"properties": set_defaults}
+        )
+
+    async def read_body(self) -> t.Tuple[str, int]:
+        """Read the body from the receive channel.
+
+        :return: A tuple (body, max_length) where max_length is the length of the larges message.
```
Comment (Ruwann) [inline connexion/validators/json.py:111]: typo: largest

**Review by Ruwann — APPROVED** [review]
I agree that the proposed validator is a good middleground :)

Perhaps also good to put this in the updated documentation, as it might otherwise be hard to find

File: connexion/validators/json.py (hunk 6 above)
Comment (RobbeSneyders) [reply connexion/validators/json.py:114]: This is not constrained by ASGI, but by the server. I ran some tests with Uvicorn and this was the upper limit, although I couldn't find where this limit is defined.
